beep-beep/jobs.md

```diff
@@ -0,0 +1,695 @@
+which allows you to dispatch tasks, and the server one, which provides the
+consumer who processes the tasks.
+
+![queue](https://user-images.githubusercontent.com/2461257/128100380-2d4df71a-c86e-4d5d-a58e-a3d503349200.png)
+
+## Installation
+
+> **Requirements**
+> - PHP >= 7.4
+> - RoadRunner >= 2.4
+> - *ext-protobuf (optional)*
+
+To get access from the PHP code, you should put the corresponding dependency
+using [the Composer](https://getcomposer.org/).
+
+```sh
+$ composer require spiral/roadrunner-jobs
+```
+
+## Configuration
+
+After installing all the required dependencies, you need to configure this
+plugin. To enable it add `jobs` section to your configuration.
+
+For example, in this way, you can configure both the client and server parts to
+work with RabbitMQ.
+
+```yaml
+#
+# RPC is required for tasks dispatching (client)
+#
+rpc:
+  listen: tcp://127.0.0.1:6001
+
+#
+# This section configures the task consumer (server)
+#
+server:
+  command: php consumer.php
+  relay: pipes
+
+#
+# AMQP connection settings (for example, RabbitMQ)
+#
+amqp:
+    addr: amqp://guest:guest@localhost:5672
+
+#
+# In this section, the jobs themselves are configured
+#
+jobs:
+  consume: [ "test" ]   # List of RoadRunner queues that can be processed by 
+                        # the consumer specified in the "server" section.
+  pipelines:
+    test:               # RoadRunner queue identifier
+      driver: amqp      # - Queue driver name
+      queue: test       # - Internal (driver's) queue identifier
+```
+
+- The `rpc` section is responsible for client settings. It is at this address
+  that we will connect, *dispatching tasks* to the queue.
+
+- The `server` section is responsible for configuring the server. Previously, we
+  have already met with its description when setting up the [PHP Worker](/php/worker.md).
+
+- The `amqp` section contains settings for connecting to RabbitMQ. As the name
+  implies, the settings are specific only for the RabbitMQ driver and *may not*
+  be available if you use some other driver.
+
+- And finally, the `jobs` section is responsible for the work of the queues
+  themselves. It contains information on how the RoadRunner should work with
+  connections to drivers, what can be handled by the consumer, and other
+  queue-specific settings.
+
+### Common Configuration
+
+Let's now focus on the common settings of the queue server. In full, it may 
+look like this:
+
+```yaml
+jobs:
+  num_pollers: 64
+  pipeline_size: 100000
+  pool:
+    num_workers: 10
+    max_jobs: 0
+    allocate_timeout: 60s
+    destroy_timeout: 60s
+  consume: [ "queue-name" ]
+  pipelines:
+    queue-name:
+      driver: # "[DRIVER_NAME]"
+      # And driver-specific configuration below...
+```
+
+Above is a complete list of all possible common Jobs settings. Let's now figure
+out what they are responsible for.
+
+- `num_pollers` - TODO
+- `pipeline_size` - TODO
```

Comment:
`pipeline_size` is used to configure the `binary heaps` priority queue. Priority queue stores jobs inside according to its' priorities. Priority might be set for the job or inherited by the pipeline.
If worker performance is poor, PQ will accumulate jobs until `pipeline_size` will be reached. After that, PQ will be blocked until workers process all the jobs inside.
Blocked PQ means, that you can `Push` the job into the driver, but RR will not read that job until PQ will be empty.
If RR will be killed with jobs inside the PQ, they won't be lost, because jobs are deleted from the drivers' queue only after `Ack`.